### ISSUE TITLE
3D Non Max Suppression

### ISSUE THREAD

**Issue by gdavid57** [tensorflow/tensorflow#48070]
**System information**

- TensorFlow version (you are using): 2.0.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

- Tensorflow incorporates a native Non Max Suppression algorithm for 2D bounding boxes.

**Will this change the current api? How?**

- The feature would add a Non Max Suppression algorithm to TF for 3D bounding boxes on the model of the 2D case algorithm.

**Who will benefit with this feature?**

- This addon will greatly benefit to the increasing community of 3D object detection or instance segmentation.

**Any Other info.**

**Comment by Saduf2019** [comment]
@gdavid57 
Please provide with more information in support of your request.

**Comment by gdavid57** [comment]
**Describe the feature and the current behavior/state.**

Tensorflow already incorporates a native Non Max Suppression algorithm for 2D bounding boxes. It is used for instance for objet detection tasks, usually after a anchor box generation step as it is observed in YOLO. Another remarkable use exists within the Mask R-CNN algorithm: many anchor boxes are generated (~200,000-300,000 boxes). After a step to reduce this number to 6,000 using some classification results, tf.image.non_max_suppression is called to sort these 6,0000 best anchors and finally obtain 2,000 regions of interest.

The feature would consist of a Non Max Suppression algorithm for three dimensional bounding boxes. This 3D Non Max Suppression would almost have the same parameters than the existing 2D implementation, called with
```
tf.image.non_max_suppression(
    boxes, scores, max_output_size, iou_threshold=0.5,
    score_threshold=float('-inf'), name=None
)
```
with boxes containing no more the 4 coordinates of the two points that define a 2D bounding box, but the 6 coordinates of the two points defining a 3D bounding box. The other parameters would stay the same. The heart of this addition would be to adapt the IOU (Intersection Over Union) function within the Non Max Suppression C file to a three dimensional IOU evaluation.

**Will this change the current api? How?**

The feature would add a Non Max Suppression algorithm to TF for 3D bounding boxes following the structure of the existing 2D implementation. It will add three files to tensorflow/tensorflow/core/kernels/image/ : non_max_suppression_3d_op.cc, non_max_suppression_3d_op.cu.cc and non_max_suppression_3d_op.h.

**Who will benefit with this feature?**

This addon will greatly benefit to the increasing community of 3D object detection or 3D instance segmentation. The demand is increasing. I give two examples: this stackoverflow (https://stackoverflow.com/questions/66311026/how-to-do-3d-non-max-suppression) alas removed by its author, and this implementation in PyTorch (https://github.com/MIC-DKFZ/medicaldetectiontoolkit) that possesses a not working 3D implementation of the NMS (the implementation author have been asked many times to correct it).

It actually exists some implementations in Python language but the speed performance given by a C implementation would be crucial when handling 3D data.

**Any Other info.**

The files to duplicate and adapt to the 3D case are mainly in the  directory: non_max_suppression_op.cc, .

**Comment by jvishnuvardhan** [comment]
@gdavid57 As this is a new feature, I think this is more suited for `tf-addons` repo [here](https://github.com/tensorflow/addons/issues). Please raise the feature request with that repo as close it here. Thanks!

**Comment by therc01** [comment]
Hi, I want to contribute to 3D NMS.

**Comment by gdavid57** [comment]
@jvishnuvardhan Alright, thank you for your response.
@therc01 Hi! follow the [way](https://github.com/tensorflow/addons/issues/2434)
